cub/cub/thread/thread_reduce.cuh

```diff
@@ -43,199 +41,515 @@
+//!
+//!   * Sum (``cub::Sum``), Bitwise AND (``cub::BitAnd``), OR (``cub::BitOr``), XOR (``cub::BitXor``) on SM50+ for
+//!     integer data types
+//!   * Minimum (``cub::Min``) and Maximum (``cub::Max``) on SM80+ for integer data types (Hopper DPX instructions),
+//!     ``__half2``, ``__nv_bfloat162``, ``__half`` (after vectorization), and ``__nv_bfloat16`` (after vectorization)
+//!     data types
+//!   * Minimum (``cub::Min``) and Maximum (``cub::Max``) on SM90+ for integer data types (Hopper DPX instructions)
+//!
+//! - Instruction-Level Parallelism (ILP) by exploiting a binary tree reduction for all other cases
+//!
+//! Simple Example
+//! ++++++++++++++++++++++++++
+//!
+//! The code snippet below illustrates a simple sum reductions over 4 integer values.
+//!
+//! .. code-block:: c++
+//!
+//!    #include <cub/cub.cuh>
+//!
+//!    __global__ void ExampleKernel(...)
+//!    {
+//!        int array[4] = {1, 2, 3, 4};
+//!        int sum      = cub::ThreadReduce(array, cub::Sum()); // sum = 10
+//!
+//! @endrst
+//!
+//! @brief Reduction over statically-sized array-like types.
+//!
+//! @tparam Input
+//!   <b>[inferred]</b> The data type to be reduced having member
+//!   <tt>operator[](int i)</tt> and must be statically-sized (size() method or static array)
+//!
+//! @tparam ReductionOp
+//!   <b>[inferred]</b> Binary reduction operator type having member
+//!   <tt>T operator()(const T &a, const T &b)</tt>
+//!
+//! @param[in] input
+//!   Array=like input
+//!
+//! @param[in] reduction_op
+//!   Binary reduction operator
+//!
+//! @return Aggregate of type <tt>cuda::std::__accumulator_t<ReductionOp, ValueT, PrefixT></tt>
+//!
+template <typename Input,
+          typename ReductionOp,
+#ifndef DOXYGEN_SHOULD_SKIP_THIS // Do not document
+          typename ValueT = ::cuda::std::__remove_cvref_t<decltype(::cuda::std::declval<Input>()[0])>,
+#else
+          typename ValueT = random_access_value_t<Input>,
+#endif // !DOXYGEN_SHOULD_SKIP_THIS
+          typename AccumT = ::cuda::std::__accumulator_t<ReductionOp, ValueT>>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE AccumT ThreadReduce(const Input& input, ReductionOp reduction_op);
+// forward declaration
+
+/***********************************************************************************************************************
+ * Internal Reduction Implementations
+ **********************************************************************************************************************/
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS // Do not document
+
+namespace detail
+{
+
+// NOTE: bit_cast cannot be always used because __half, __nv_bfloat16, etc. are not trivially copyable
+template <typename Output, typename Input>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE Output unsafe_bitcast(const Input& input)
+{
+  Output output;
+  static_assert(sizeof(input) == sizeof(output), "wrong size");
+  ::memcpy(&output, &input, sizeof(input));
+  return output;
+}
```

Comment:
Is that still an issue?

Reply:
yes, unfortunately. The problem is with array of extended floating-point types.